scopesim/optics/image_plane_utils.py

```diff
@@ -155,60 +164,52 @@ def header_from_list_of_xy(x, y, pixel_scale, wcs_suffix=""):
     hdr : fits.Header
 
     """
+    # TODO: find out how this plays with chunks
     s = wcs_suffix
-    if wcs_suffix != "D":
-        x = np.array(x)
-        x[x > 270] -= 360
-        x[x <= -90] += 360
-        x = list(x)
+    if wcs_suffix != "D" and not arcsec:
+        x = _fix_360(x)
+        y = _fix_360(y)
+    pnts = np.array([x, y])
 
-    hdr = fits.Header()
+    # TODO: test whether abs(pixel_scale) breaks anything
+    pixel_scale = abs(pixel_scale)
+    extent = pnts.ptp(axis=1) / pixel_scale
+    naxis = extent.round().astype(int)
 
-    # .. todo: find out how this plays with chunks
-    # crval1 = divmod(min(x), pixel_scale)[0] * pixel_scale
-    # crval2 = divmod(min(y), pixel_scale)[0] * pixel_scale
+    if s == "D":
+        crpix = np.array([1., 1.])
+        crval = pnts.min(axis=1)
+    else:
+        crpix = (naxis + 1) / 2
+        crval = (pnts.min(axis=1) + pnts.max(axis=1)) / 2
 
-    # naxis1 = int((max(x) - crval1) // pixel_scale) # + 2
-    # naxis2 = int((max(y) - crval2) // pixel_scale) # + 2
+    # To deal with half pixels:
+    offset = 0.5 * pixel_scale if s == "D" or sky_offset else 0.0
 
-    crval1 = min(x)
-    crval2 = min(y)
+    ctype = "LINEAR" if s in "DX" else "RA---TAN"
+    if s == "D":
+        cunit = "mm"
+    elif arcsec:
+        cunit = "arcsec"
+    else:
+        cunit = "deg"
 
-    # ..todo:: test whether abs(pixel_scale) breaks anything
-    pixel_scale = abs(pixel_scale)
-    dx = (max(x) - min(x)) / pixel_scale
-    dy = (max(y) - min(y)) / pixel_scale
-    naxis1 = int(np.round(dx))
-    naxis2 = int(np.round(dy))
+    new_wcs = WCS(key=s)
+    new_wcs.wcs.ctype = 2 * [ctype]
+    new_wcs.wcs.cunit = 2 * [cunit]
+    new_wcs.wcs.cdelt = np.array(2 * [pixel_scale])
+    new_wcs.wcs.crval = crval + offset
+    new_wcs.wcs.crpix = crpix
 
+    hdr = fits.Header()
     hdr["NAXIS"] = 2
-    hdr["NAXIS1"] = naxis1
-    hdr["NAXIS2"] = naxis2
-    hdr["CTYPE1"+s] = "LINEAR" if s == "D" else "RA---TAN"
-    hdr["CTYPE2"+s] = "LINEAR" if s == "D" else "DEC--TAN"
-    hdr["CUNIT1"+s] = "mm" if s == "D" else "deg"
-    hdr["CUNIT2"+s] = "mm" if s == "D" else "deg"
-    hdr["CDELT1"+s] = pixel_scale
-    hdr["CDELT2"+s] = pixel_scale
-    hdr["CRVAL1"+s] = crval1
-    hdr["CRVAL2"+s] = crval2
-    hdr["CRPIX1"+s] = 0.
-    hdr["CRPIX2"+s] = 0.
-
-    xpcen, ypcen = naxis1 // 2, naxis2 // 2
-    xscen, yscen = pix2val(hdr, xpcen, ypcen, s)
-    hdr["CRVAL1"+s] = float(xscen)
-    hdr["CRVAL2"+s] = float(yscen)
-    hdr["CRPIX1"+s] = xpcen
-    hdr["CRPIX2"+s] = ypcen
+    hdr["NAXIS1"] = naxis[0]
+    hdr["NAXIS2"] = naxis[1]
```

Comment:
Well, those where interesting hours of my life. But apparently this makes `NAXIS1` and `2` `numpy.int64`, which somehow confuses something else down the line. Making these `int` will 'solve' the problem.

To reproduce the problem you can simulate a star and save the fits file and look at the headers. Or you can run the simplestars notebook in #273 .

```suggestion
    hdr["NAXIS1"] = int(naxis[0])
    hdr["NAXIS2"] = int(naxis[1])
```

But this isn't a proper fix, because this should not matter at all.

Reply:
Now how I found this: I created a version of `image_plane_utils.py` that had both the old and the new `header_from_list_of_xy()` function, called `header_from_list_of_xy_old()` and `header_from_list_of_xy_new()`, and the original something like:
```python
def header_from_list_of_xy(...):
    hdr_old = header_from_list_of_xy_old(...)
    hdr_new = header_from_list_of_xy_new(...)
   assert hdr_old == hdr_new
   return hdr_new
```
And then I messed with the new function until the assert passed, so the output should have been exactly the same. Things I changed in order to make them equal:
- Removed "WCSAXESD", "WCSAXES", "LATPOLE", "LATPOLED", "MJDREF",
- Removed comments that WCS adds,
- Replaced values that where almost identical, like -5.4210108624275e-20 and -5.421010862427522e-20, and finally
- Reordered the header keywords.

I even wrote the old and the new headers to file and compared the hashes of those. But the headers returned by the new function would just not work, even with everything seemingly equal. I even looked whether either the new or the old function modified some global variable somehow.

But it was the types...

I'll turn the notebook of #273 into a proper test to ensure this keeps work. Actually that was my goal for this evening, but well..

You must have some similar horror stories of the last week.

Reply:
Huge thanks for investigating this!! It would have taken me a few days probably...

I'm perfectly fine with your "solution" of just converting to `int`. What I wasn't happy about in the first place, was that this manual writing to the header is at all necessary. All other parts come to the header via `WCS.to_header()`, after being passed to the `WCS` object (which does some checks and modifications according to the docs). But I couldn't figure out how to equip a `WCS` created from scratch with the necessary `NAXISi`. Which is weird, because when you create a `WCS` from a header that has those, they do become an internal attribute, somehow (and show up in the `__repr__`, for example).

In a complete `fits.ImageHDU` object, the `NAXISi` keywords are based on the shape of the `.data`. Even if you create an otherwise empty `ImageHDU` without a header and using a simple 2D array for the data, the HDU will generate a basic header with those keywords, and adapt them when the shape of `.data` is mutated. What I'm saying is, maybe we don't (or shouldn't) need these keywords in headers created with this function. Or maybe the function should create WCSs and not headers. Or maybe Manuel Neuer. I should go to bed 😂 

And finally, yes, the past 10 days or so of bug fixing were indeed one hell of a ride at times 🙃